pkg/lib/cockpit/channel.ts

```diff
@@ -0,0 +1,240 @@
+    _payload_type: P | null = null; // required to forbid mixing types
+
+    id: string | null = null; // can be unassigned during transport startup
+    readonly options: ChannelOptions<P>;
+    readonly binary: ChannelBinaryOption<P>['binary'];
+
+    #transport: Transport | null = null;
+    #received: Partial<Record<"close" | "ready" | "done", ChannelControlMessage>> = {};
+    #queue: ([true, ChannelControlMessage] | [false, P])[] = [];
+
+    on_control(control: ChannelControlMessage) {
+        const command = control.command;
+
+        if (command === 'ready' || command === 'close' || command === 'done') {
+            if (command in this.#received) {
+                return;
+            }
+
+            this.#received[command] = control;
+            this.emit(command, control);
+        } else {
+            this.emit('control', control);
+        }
+
+        if (command === 'close') {
+            if (this.#transport && this.id)
+                this.#transport.unregister(this.id);
+            if (control.message)
+                console.warn(control.message);
+        }
+    }
+
+    constructor(options: ChannelOptions<P>) {
+        super();
+
+        this.options = { ...options };
+        this.binary = (options.binary === true);
+
+        ensure_transport(transport => {
+            if ('close' in this.#received)
+                return;
+
+            this.#transport = transport;
+            this.id = transport.next_channel();
+            transport.register(this.id, control => this.on_control(control), data => this.emit('data', data));
+
+            // We need to delay sending the open message until after we have
+            // the transport because we need to set the host field.
+
+            // Make a copy so we can modify some fields.
+            const command: JsonObject = { ...this.options };
+
+            if (!('host' in command) && transport_globals.default_host) {
+                command.host = transport_globals.default_host;
+            }
+
+            if (this.binary) {
+                command.binary = "raw";
+            } else {
+                delete command.binary;
+            }
+
+            // Go direct: we need this to go before the rest of the queue
+            transport.send_control({ ...command, command: 'open', channel: this.id, 'flow-control': true });
+
+            // Now send everything else from the queue
+            for (const [is_control, message] of this.#queue) {
+                if (is_control) {
+                    transport.send_control({ ...message, channel: this.id });
+                } else {
+                    transport.send_message(message, this.id);
+                }
+            }
+            this.#queue = [];
+        });
+    }
+
+    /**
+     * Sends a payload frame.
+     *
+     * @message the payload to send, either a string or a Uint8Array.
+     */
+    send_data(message: P) {
```

Comment:
pkg/lib/_internal/channel.js warns if you try to send data on a closed channel. This seems like a good idea?

Reply:
I removed some of the warnings because it made the state tracking annoying.  I can add this one back.

Reply:
I put them back.

One interesting and potentially annoying aspect of this, though: close can come from the server side asynchronous to sending.  That means that in order not to violate this rule you need to register a close handler and stop sending once close happens.